File: mirea/cources/2026-2027/aii-july/01/Part-04-Binary-search/examples/example-14.py
# Максимальный элемент, меньший x, в Python3

from bisect import bisect_left

# Чтение отсортированного массива и значения x
n = int(input())
a = list(map(int, input().split()))
x = int(input())

# Ручной бинарный поиск последнего индекса, на котором a[index] < x
left = -1
right = n
while right - left > 1:
    mid = (left + right) // 2
    if a[mid] < x:
        left = mid
    else:
        right = mid

# Через bisect_left достаточно взять предыдущий элемент
pos = bisect_left(a, x)
has_manual = left != -1
has_library = pos != 0

if has_manual != has_library:
    raise RuntimeError("Ответы не совпали")
if has_manual and a[left] != a[pos - 1]:
    raise RuntimeError("Ответы не совпали")

# Выводим ответ двумя способами
if not has_manual:
    print("NO")
    print("NO")
else:
    print(a[left])
    print(a[pos - 1])

File: mirea/cources/2026-2027/aii-july/01/Part-04-Binary-search/examples/example-14.cpp
// Максимальный элемент, меньший x, в C++

#include <algorithm>
#include <iostream>
#include <iterator>
#include <vector>

using namespace std;

int main() {
    // Чтение отсортированного массива и значения x
    int n;
    cin >> n;

    vector<int> a(n);
    for (int i = 0; i < n; i++) {
        cin >> a[i];
    }

    int x;
    cin >> x;

    // Ручной бинарный поиск последнего индекса, на котором a[index] < x
    int left = -1;
    int right = n;
    while (right - left > 1) {
        int mid = left + (right - left) / 2;
        if (a[mid] < x) {
            left = mid;
        } else {
            right = mid;
        }
    }

    // Через lower_bound достаточно взять предыдущий элемент
    auto it = lower_bound(a.begin(), a.end(), x);
    bool has_manual = left != -1;
    bool has_library = it != a.begin();

    if (has_manual != has_library) {
        return 1;
    }
    if (has_manual && a[left] != *prev(it)) {
        return 1;
    }

    // Выводим ответ двумя способами
    if (!has_manual) {
        cout << "NO\n";
        cout << "NO\n";
    } else {
        cout << a[left] << '\n';
        cout << *prev(it) << '\n';
    }
}